shared/src/main/scala/coop/rchain/shared/ByteStringOps.scala

```diff
@@ -0,0 +1,116 @@
+package coop.rchain.shared
+
+import java.io._
+import java.util.concurrent.{Callable, Executors}
+
+import com.google.protobuf.ByteString
+import org.apache.commons.compress.compressors.lz4.FramedLZ4CompressorOutputStream.Parameters
+import org.apache.commons.compress.compressors.lz4.{
+  FramedLZ4CompressorInputStream,
+  FramedLZ4CompressorOutputStream
+}
+
+import scala.util.{Failure, Success, Try}
+
+object ByteStringOps {
+
+  implicit class RichByteString(bs: ByteString) {
+    def compress: ByteString = compressLZ4(bs)
+
+    def decompress: Option[ByteString] = decompressLZ4(bs)
+  }
+
+  private def compressLZ4(bs: ByteString): ByteString = {
+    val pos = new PipedOutputStream()
+    val pis = new PipedInputStream(pos)
+    val is  = bs.newInput()
+
+    // Writes compressed data into `pos`
+    val producer = new Runnable {
+      override def run(): Unit = {
+        var lz4os: FramedLZ4CompressorOutputStream = null
+        try {
+          lz4os = new FramedLZ4CompressorOutputStream(pos, Parameters.DEFAULT)
+          val data  = new Array[Byte](8192)
+          var count = is.read(data)
+
+          while (count != -1) {
+            lz4os.write(data, 0, count)
+            count = is.read(data)
+          }
```

Comment:
this can be extracted into a `transferAll(InputStream, OutputStream)` function and used twice